include/oneapi/dpl/pstl/hetero/dpcpp/sycl_iterator.h

```diff
@@ -131,25 +131,44 @@ struct sycl_iterator
     }
 };
 
-// mode converter when property::noinit present
-template <access_mode __mode>
-struct _ModeConverter
+// map access_mode tag to access_mode value
+template <typename _ModTagT, typename _NoInitT = void>
+struct __access_mode_resolver
 {
-    static constexpr access_mode __value = __mode;
 };
 
 template <>
-struct _ModeConverter<access_mode::read_write>
+struct __access_mode_resolver<std::decay_t<decltype(sycl::read_only)>, void>
 {
-    static constexpr access_mode __value = access_mode::discard_read_write;
+    static constexpr access_mode __mode = access_mode::read;
+};
+
+template <>
+struct __access_mode_resolver<std::decay_t<decltype(sycl::write_only)>, void>
```

Comment:
Should this `std::decay_t` be here?

If `decltype(sycl::write_only)` is `const`, won't this no longer match with someone passing in `sycl::write_only{}` to the public api?  

I might be wrong, but either I think we should either not use decay here, or probably even better, we should use decay also below when using `__access_mode_resolver`, or in some implementation layer automatically on the incoming type.

Reply:
I spent quite a lot of time to get to `std::decay_t`, originally having decay-less specializations. Here is an example to play with, where the types do not match: https://godbolt.org/z/58Edjs4fr. You can uncomment lines 40-41 and check the error message. This puzzled me, and you may get surprised too. 

> If decltype(sycl::write_only) is const, won't this no longer match with someone passing in sycl::write_only{} to the public api?

I guess this is the exact problem. These tags are `constexpr`, which assumes `const`. 

> we should either not use decay here, or probably even better, we should use decay also below

I will do so. I am not sure if decay is needed when we call `__access_mode_resolver`, as the specialization without `const` is more permissive, but it will not hurt.

Reply:
You are right, I am surprised, I need to dig into this better because there is something I must not understand.  I think it is about the types in the template function instantiation.

Reply:
Ah, ok I understand now, I believe it is simply that it loses its `const` qualifier when it instantiates the template function.  
Passing a `const` qualified argument by value into a template function does not force that function to treat that variable as a `const` type.  It is the responsibility of the writer of the template function to figure out if that is a const param or not.  

I think your implementation is fine, but it may prevent any future confusion to decay on the way in to `__access_mode_resolver`, or alternatively, mark `ModeTagT` const and name the parameter `tag`, then take `decltype(tag)` on the way in the `__access_mode_resolver`.   

Really any of these options works for me.  Thanks for indulging me in my confusion.